File: exercises_test.py
from exercises import *
import pytest

@pytest.mark.skip(reason="Delete this once you've implemented the Stack class")
def test_stack():
    stack = Stack()
    assert stack.is_empty()
    stack.push('a')
    assert not stack.is_empty()
    assert stack.pop() == 'a'
    assert stack.is_empty()
    stack.push('b')
    stack.push('c')
    assert stack.peek() == 'c'
    assert stack.size() == 2
    assert stack.pop() == 'c'

@pytest.mark.skip(reason="Delete this once you've implemented the Queue class")
def test_queue():
    queue = Queue()
    assert queue.is_empty()
    queue.enqueue('1st')
    assert not queue.is_empty()
    queue.enqueue('2nd')
    queue.enqueue('3rd')
    assert queue.size() == 3
    assert queue.dequeue() == '1st'
    assert queue.size() == 2

@pytest.mark.skip(reason="Delete this once you've implemented the Deque class")
def test_deque():
    deque = Deque()
    assert deque.is_empty()
    deque.add_rear(4)
    deque.add_rear('Dog')
    deque.add_front('Cat')
    deque.add_front(True)
    assert deque.size() == 4
    assert not deque.is_empty()
    deque.add_rear(8.4)
    assert deque.remove_rear() == 8.4
    assert deque.remove_front() == True

@pytest.mark.skip(reason="Delete this once implemented")
def test_is_balanced_parentheses():
    assert is_balanced_parentheses('()()') == True
    assert is_balanced_parentheses('()((') == False
    assert is_balanced_parentheses('((()))()') == True
    assert is_balanced_parentheses(')()(') == False

@pytest.mark.skip(reason="Delete this once implemented")
def test_is_palindrome():
    assert is_palindrome('abba')
    assert is_palindrome('racecar')
    assert not is_palindrome('palindrome')
    assert not is_palindrome('08052020')
    assert is_palindrom('02022020')

@pytest.mark.skip(reason="Delete this once implemented")
def test_decimal_to_binary():
    assert decimal_to_binary(7) == '111'
    assert decimal_to_binary(42) == '101010'
    assert decimal_to_binary(225) == '11100001'

File: exercises.test.js
const { Stack, Queue, Deque, isBalancedParentheses, isPalindrome, decimalToBinary } =  require('./exercises.js');

describe('Stack', () => {
  it('is empty by default', () => {
    const stack = new Stack();
    expect(stack.isEmpty()).toEqual(true);
  });

  it('can add items', () => {
    const stack = new Stack();
    stack.push('a');
    expect(stack.isEmpty()).toEqual(false);
  });

  it('can remove items', () => {
    const stack = new Stack();
    stack.push('a');
    expect(stack.pop()).toEqual('a');
    expect(stack.isEmpty()).toEqual(true);
  });

  it('can hold mulitiple items', () => {
    const stack = new Stack();
    stack.push('b');
    stack.push('c');
    expect(stack.size()).toEqual(2);
  });

  it('can preview the next item', () => {
    const stack = new Stack();
    stack.push('b');
    stack.push('c');
    expect(stack.peek()).toEqual('c');
  });
});

describe('Queue', () => {
  it('is empty by default', () => {
    const queue = new Queue();
    expect(queue.isEmpty()).toEqual(true);
  });

  it('can add items', () => {
    const queue = new Queue();
    queue.enqueue('1st');
    expect(queue.isEmpty()).toEqual(false);
  });

  it('can add multiple items', () => {
    const queue = new Queue();
    queue.enqueue('1st');
    queue.enqueue('2nd');
    queue.enqueue('3rd');
    expect(queue.size()).toEqual(3);
  });

  it('can remove items', () => {
    const queue = new Queue();
    queue.enqueue('1st');
    queue.enqueue('2nd');
    queue.enqueue('3rd');
    expect(queue.dequeue()).toEqual('1st');
    expect(queue.size()).toEqual(2);
  });
});

describe('Deque', () => {
  it('is empty by default', () => {
    const deque = new Deque();
    expect(deque.isEmpty()).toEqual(true);
  });

  it('can add to the rear and front', () => {
    const deque = new Deque();
    deque.addRear(4);
    deque.addRear('Dog');
    deque.addFront('Cat');
    deque.addFront(true);
    expect(deque.size()).toEqual(4);
    expect(deque.isEmpty()).toEqual(false);
    deque.addRear(8.4);
    expect(deque.removeRear()).toEqual(8.4);
    expect(deque.removeFront()).toEqual(true);
  });
});

describe('isBalancedParentheses', () => {
  it('can determine parentheses balance', () => {
    expect(isBalancedParentheses('()()')).toEqual(true);
    expect(isBalancedParentheses('()((')).toEqual(false);
    expect(isBalancedParentheses('()(()))')).toEqual(false);
    expect(isBalancedParentheses('((()))()')).toEqual(true);
    expect(isBalancedParentheses(')()(')).toEqual(false);
  })
});

describe('isPalindrome', () => {
  it('can determine if a string is a palindrome', () => {
    expect(isPalindrome('abba')).toEqual(true);
    expect(isPalindrome('racecar')).toEqual(true);
    expect(isPalindrome('palindrome')).toEqual(false);
    expect(isPalindrome('08052020')).toEqual(false);
    expect(isPalindrome('02022020')).toEqual(true);
  });
});

describe('decimalToBinary', () => {
    it('can convert decmials to binary', () => {
      expect(decimalToBinary(7)).toEqual('111');
      expect(decimalToBinary(42)).toEqual('101010');
      expect(decimalToBinary(225)).toEqual('11100001');
    });
});
